makefiles/boot/riotboot.mk

```diff
@@ -0,0 +1,122 @@
+ifneq (,$(filter riotboot,$(FEATURES_USED)))
+
+.PHONY: riotboot/flash riotboot/flash-bootloader riotboot/flash-slot0 riotboot/bootloader/%
+
+RIOTBOOT_DIR = $(RIOTBASE)/bootloaders/riotboot
+RIOTBOOT ?= $(RIOTBOOT_DIR)/bin/$(BOARD)/riotboot.elf
+CFLAGS += -I$(BINDIR)/riotbuild
+
+HEADER_TOOL_DIR = $(RIOTBASE)/dist/tools/riotboot_gen_hdr
+HEADER_TOOL ?= $(HEADER_TOOL_DIR)/bin/genhdr
+BINDIR_APP = $(BINDIR)/$(APPLICATION)
+
+# Indicate the reserved space for a header, 256B by default
+# Notice that it must be 256B aligned. This is restricted by
+# the Cortex-M0+/3/4/7 architecture
+RIOTBOOT_HDR_LEN ?= 0x100
+
+# Export variables for 'riotboot_slot'
+export SLOT0_LEN
+
+# By default, slot 0 is found just after RIOTBOOT_LEN. It might
+# be overridden to add more offset as needed.
+export SLOT0_OFFSET ?= $(RIOTBOOT_LEN)
+
+# Mandatory APP_VER, set to 0 by default
+APP_VER ?= 0
+
+# Final target for slot 0 with riot_hdr
+SLOT0_RIOT_BIN = $(BINDIR_APP)-slot0.riot.bin
+
+# For slot generation only link is needed
+$(BINDIR_APP)-%.elf: link
+	$(Q)$(_LINK) -o $@
+
+# Slot 0 firmware offset, after header
+SLOT0_IMAGE_OFFSET := $$(($(RIOTBOOT_LEN) + $(RIOTBOOT_HDR_LEN)))
+
+# Link slots ELF *after* riot_hdr and limit the ROM to the slots length
+$(BINDIR_APP)-slot0.elf: FW_ROM_LEN=$$((SLOT0_LEN - $(RIOTBOOT_HDR_LEN)))
+$(BINDIR_APP)-slot0.elf: ROM_OFFSET=$(SLOT0_IMAGE_OFFSET)
+
+# Create binary target with RIOT header
+$(SLOT0_RIOT_BIN): %.riot.bin: %.hdr %.bin
+	@echo "creating $@..."
+	$(Q)cat $^ > $@
+
+# Compile header tool if it doesn't exist, force its compilation in case
+# some files changed
+$(HEADER_TOOL): FORCE
+	@echo "compiling $@..."
+	$(Q)/usr/bin/env -i \
+		QUIET=$(QUIET) \
+		PATH=$(PATH) \
+			$(MAKE) --no-print-directory -C $(HEADER_TOOL_DIR) all
+
+# Generate RIOT header and keep the original binary file
+# It must be always regenerated in case of any changes, so FORCE
+.PRECIOUS: %.bin
```

Comment:
Why the .PRECIOUS? Ideally, if you need the .bin for some rule you should state that explicitly.

Reply:
It's more for debugging purposes. The intermediate bin files are not really needed, but provide information about its generation which helps to know if they are correct. On the other hand, they might also be distributed separately. This is mostly an optional thing which does't affect neither the compilation process (speed) or the size of the build (their size is just a few bytes).

Reply:
Ok, undesrtood. Again, maybe it's good to document it so that we know that it is there for debugging only. Even then, I think that if the bin is wanted it should be explicitly made a dependency of riotboot or something like that.